File: SocialJournal/SocialJournal/MapViewController.m
//
//  MapViewController.m
//  SocialJournal
//
//  Created by Matt Phillips on 11/12/14.
//  Copyright (c) 2014 UH. All rights reserved.
//

#import "MapViewController.h"

@interface MapViewController ()
@property (weak, nonatomic) IBOutlet MKMapView *mapView;
@property CLPlacemark *placmark;

@end

@implementation MapViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self getNearestCity:29.76 :-95.36];
    [self showMapLocation:29.76 :-95.36];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}
- (void) showMapLocation:(double)latitude :(double)longitude {
    
    CLLocationCoordinate2D location;
    location.latitude = latitude;
    location.longitude = longitude;
    MKPointAnnotation *annotation = [[MKPointAnnotation alloc] init];
    annotation.coordinate = location;
    [self.mapView addAnnotation:annotation];
    //[self.mapView setShowsUserLocation:YES];
}

-(NSString*)getNearestCity :(double)latitude :(double)longitude{
    __block NSString *result;
    CLGeocoder *ceo = [[CLGeocoder alloc]init];
    CLLocation *loc = [[CLLocation alloc]initWithLatitude:latitude longitude:longitude];
    
    [ceo reverseGeocodeLocation: loc completionHandler:
     ^(NSArray *placemarks, NSError *error) {
         CLPlacemark *placemark = [placemarks objectAtIndex:0];
         result = placemark.locality; // Extract the city name
        }
     ];
    return result;
}
/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: SocialJournalSwift/SocialJournal/SocialJournal/MapViewController.swift
//
//  MapViewController.swift
//  SocialJournal
//
//  Created by Matt Phillips on 11/20/14.
//  Copyright (c) 2014 UH. All rights reserved.
//

import Foundation
import MapKit

class MapViewController: UIViewController, MKMapViewDelegate{
    
    @IBOutlet var mapView: MKMapView!
    var currentEntry = PFObject(className: "Entry")
    
    override func viewDidLoad() {
        super.viewDidLoad()
            
        var location = CLLocation(latitude: self.currentEntry["location"].latitude, longitude: self.currentEntry["location"].longitude)
        
        var span = MKCoordinateSpanMake(0.5, 0.5)
        var region = MKCoordinateRegion(center: location.coordinate, span: span)
            
        self.mapView.setRegion(region, animated: true)
            
        var annotation = MKPointAnnotation()
        annotation.setCoordinate(location.coordinate)
        annotation.title = "city/locality"
        annotation.subtitle = "region"
        
        CLGeocoder().reverseGeocodeLocation(location, completionHandler: {(placemarks, error)->Void in
            if placemarks.count > 0 {
                let pm = placemarks[0] as CLPlacemark
                annotation.title = pm.locality
                annotation.subtitle = pm.country
                //println("Placemark: " + pm.locality + "Annotation: " + annotation.title)
                //println("Placemark: " + pm.country + "Annotation: " + annotation.subtitle)
            }
        })
        self.mapView.addAnnotation(annotation)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func mapView(mapView: MKMapView!, viewForAnnotation annotation: MKAnnotation!) -> MKAnnotationView! {
        if !(annotation is MKPointAnnotation) {
            //if annotation is not an MKPointAnnotation (eg. MKUserLocation),
            //return nil so map draws default view for it (eg. blue dot)...
            return nil
        }
        
        let reuseId = "test"
        var anView = mapView.dequeueReusableAnnotationViewWithIdentifier(reuseId)
        if anView == nil {
            anView = MKAnnotationView(annotation: annotation, reuseIdentifier: reuseId)
            anView.canShowCallout = true
        }
        else {
            //we are re-using a view, update its annotation reference...
            anView.annotation = annotation
        }
        
        return anView
    }
    
    
}